### PR TITLE
Stub testing support

### PR DESCRIPTION
**Pull request by chinmaydd**
### Description of changes: 

This PR adds support for testing stub implementations through compiletest. It also adds an initial set of tests for the `RMCVec` abstraction mostly extracted from the Rust reference and adapted for RMC.

Example:
```bash
$ ./x.py test -i --stage 1 stub-tests
```

### Resolved issues:

#372

### Call-outs:

NOTE: This PR depends on #377.

### Testing:

* How is this change tested?
This change is specifically for testing the stub implementations.

* Is this a refactor change?
No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/test/benchmarks/Vector/append.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    let mut v2: Vec<u32> = Vec::with_capacity(times);
+    for i in 0..times {
+        v2.push(__nondet());
+    }
+    v.append(&mut v2);
+    assert!(v.len() == times * 2);
+}
+
+fn main() {
+    operate_on_vec(100);
```
Comment (danielsn) [inline src/test/benchmarks/Vector/append.rs:23]: do you want to make this a parameter

File: src/test/benchmarks/Vector/extend.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+include!{"../../rmc-prelude.rs"}
+
+fn operate_on_vec(times: usize) {
+    let mut v = Vec::with_capacity(times);
+    v.extend(1..times);
+}
+
+fn main() {
+    operate_on_vec(100);
```
Comment (danielsn) [inline src/test/benchmarks/Vector/extend.rs:15]: Maybe have `benchmark_prelude.rs`

File: src/test/benchmarks/Vector/insert.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    let mut v: Vec<u32> = Vec::with_capacity(times);
+    for i in 0..times {
+        v.push(__nondet());
+    }
+    // Here the Vecs will grow internally and perform memmove to accomodate
+    // new element.
+    // for i in 0..times {
+    v.insert(v.len()/2, __nondet());
+    // }
+    v.push(5);
```
Comment (danielsn) [inline src/test/benchmarks/Vector/insert.rs:19]: If you put this before the `v.insert()` it would be a more interresting test

Comment (danielsn) [reply src/test/benchmarks/Vector/insert.rs:19]: Also, you could make it slightly more genereral:
```rust
let sentinal = __nondet();
v.push(sentinal);
...
assert!(v.pop() == Some(sentinal));
```

File: src/test/benchmarks/Vector/remove.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    let mut v: Vec<u32> = Vec::with_capacity(times);
+    for i in 0..times {
+        v.push(__nondet());
+    }
+    // We remove elements to perform more memmoves. These are done to fill up
+    // "holes" created due to elements removed from the middle of the Vec.
+    for i in 0..times/2 {
+        v.remove(v.len()/2);
+    }
+    v.push(5);
```
Comment (danielsn) [inline src/test/benchmarks/Vector/remove.rs:19]: put this earlier

File: src/test/benchmarks/Vector/remove.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    assert!(v.pop() == Some(5));
```
Comment (danielsn) [inline src/test/benchmarks/Vector/remove.rs:20]: assert vec length

File: src/test/benchmarks/Vector/shrink_and_clear.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    for i in 0..times {
+        v.push(__nondet());
+    }
+    assert!(v.len() == times);
+    // Here, Vecs with grow() internally
+    v.push(1);
+    assert!(v.pop() == Some(1));
+    // Completely shrink the array to remove additional allocations
+    v.shrink_to_fit();
+    // Push some new elements to grow() again
```
Comment (danielsn) [inline src/test/benchmarks/Vector/shrink_and_clear.rs:20]: now check that the last element is right. Even better: take a nondet i, save its value, then check it hasn't changed

File: src/test/stub-tests/HashSet/main.rs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// rmc-flags: --use-abs --abs-type c-ffi
+include!{"../../rmc-prelude.rs"}
+
+fn main() {
+    let mut h: HashSet<u16> = HashSet::new();
+
+    assert!(h.insert(10));
```
Comment (danielsn) [inline src/test/stub-tests/HashSet/main.rs:10]: Use nondet values here would be neater

Comment (danielsn) [reply src/test/stub-tests/HashSet/main.rs:10]:
```rust
let a = __nondet();
let b = __nondet();
__VERIFIER_assume(a != b);
```

File: src/test/stub-tests/Vec/append.rs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (danielsn) [inline src/test/stub-tests/Vec/append.rs:1]: This is verbabim from the rust reference

File: src/test/stub-tests/HashSet/main.rs (hunk 7 above)
Comment (danielsn) [reply src/test/stub-tests/HashSet/main.rs]: Make another version of the test, mark skip, file ticket for known performance issue

**Review by danielsn — APPROVED** [review]
Approved, modulo our discussed changes

File: src/test/stub-tests/HashSet/main.rs (hunk 7 above)
Comment (chinmaydd) [reply src/test/stub-tests/HashSet/main.rs]: Created issue: https://github.com/model-checking/rmc/issues/478 for tracking

**Pull request merged**